catalyst/data/sampler.py

```diff
@@ -76,6 +81,99 @@ def __len__(self) -> int:
         return self.length
 
 
+class BalancecBatchSampler(Sampler):
+    """
+    Sampler with the given strategy for the dataset with C unique classes:
+    - Selection P of C classes for the 1st batch
+    - Selection K instances for each class for the 1st batch
+    - Selection P of C - P remaining classes for 2nd batch
+    - Selection K instances for each class for the 2nd batch
+    - ...
+    The epoch ends when there are no classes left.
+    So, the batch sise is P * K except the last one.
+
+    Thus, in each epoch, all the classes will be selected once, but this
+    does not mean that all the instances will be selected during the epoch.
+
+    One of the purposes of this sampler is to be used for
+    forming triplets and pos/neg pairs inside the batch.
+    To guarante existance of these pairs in the batch,
+    P and K should be > 1. (1)
+
+    Behavior in corner cases:
+    - If a class does not contain K instances, a choice will be made with repetition.
+    - If C % P == 1 then one of the classes should be dropped otherwise
+      statement (1) will not be met.
+
+    This type of sampling can be found in the classical paper of Person Re-Id:
+    "In Defense of the Triplet Loss for Person Re-Identification",
+    where P equals 32 and K equals 4.
+    """
+
+    def __init__(self, labels: List[int], p: int, k: int):
+        """
+
+        Args:
+            labels: list of classes labeles for each elem in the dataset
+            p: number of classes in a batch, should be > 1
+            k: number of instances of each class in a batch, should be > 1
+        """
+
+        super().__init__(self)
+        classes = set(labels)
+
+        assert isinstance(p, int) and isinstance(k, int)
+        assert (1 < p <= len(classes)) and (1 < k)
+        assert all(
+            [n > 1 for n in Counter(labels).values()]
+        ), "Each class shoud contain at least 2 instances to fit (1)"
+
+        self._labels = tuple(labels)  # to make it immutable
+        self._p = p
+        self._k = k
+
+        self._bs = self._p * self._k
+        self._classes = classes
+
+        # to satisfy statement (1)
+        n_classes = len(self._classes)
```

Comment:
`num_classes ` maybe? it would be more appropriate for Catalyst codestyle

Reply:
done